Add new ReportFilters component

ryelle opened the pull request:
This is a new component which serves as a collection for all the filter components available on a report. It's basically a way to centralize adding & styling these filters.

This is still in progress - placeholder code in a lot of places - but I have a few questions so this seemed the easiest way to share.

- Should all the filters move into this folder? I plan to move `analytics/filters` to `components/report-filters`, and currently the other filters only show up as part of report-filters

ryelle commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,84 @@
+
+const ReportFilters = ( { advancedConfig, filters, filterPaths, query, path } ) => {
+	let advancedCard = false;
+	switch ( query.filter ) {
```
We have `FILTER_PARAM` defined in `filter-picker`, but it should always be `filter` - I think it would be more clear to use `query.filter` directly. 

If we _do_ need to use constants for query params, maybe we should collect them somewhere (in the nav-utils library?) to make sure we're not overlapping anywhere?

ryelle commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,84 @@
+		case 'advanced':
+			advancedCard = (
+				<AdvancedFilters config={ advancedConfig } filterTitle={ __( 'Orders', 'wc-admin' ) } />
+			);
```
The above is partially placeholder, but it does bring up the same label issues that `AdvancedFilters` has, where we want to be able to pass around labels as simply as possible. See #285

ryelle commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,84 @@
+					{ !! filters.length && (
+						<FilterPicker
+							query={ query }
+							path={ path }
```
Do we need to pass through `path`/`query`? Since this should be based on the current page, can each picker [access these from `history.location`](https://github.com/woocommerce/wc-admin/blob/master/client/lib/nav-utils.js#L23-L24) (maybe via helper function in `lib/nav-utils`)?

ryelle commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,84 @@
+							query={ query }
+							path={ path }
+							filters={ filters }
+							filterPaths={ filterPaths }
```
Could filters + filterPaths merge into one config object?

psealock replied:
sure

ryelle commented:
@woocommerce/isotope (and specifically @psealock ) Can I get some feedback on the questions I've asked above ^? Thanks!

psealock commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,85 @@
+					) }
+				</div>
+				{ false !== advancedCard && (
+					<div className="woocommerce-analytics__advanced-filters">{ advancedCard }</div>
```
Is this div needed?

ryelle replied:
Both divs here are "row containers", so we can style things easier.

psealock commented on client/analytics/filters/index.js:
```diff
@@ -0,0 +1,85 @@
+					<div className="woocommerce-analytics__advanced-filters">{ advancedCard }</div>
+				) }
+			</Section>
+		</Fragment>
```
The Product Detail page introduces a third filter for selecting a particular product: p6riRB-3dP-p2

![screen shot 2018-08-13 at 11 56 28 am](https://user-images.githubusercontent.com/1922453/44007847-388a25ee-9ef0-11e8-94ef-c5b69f17237d.png)

Logic and CSS will need to live here as well.

ryelle replied:
Should be easy enough to drop in once we're done, and we can again pass in conditional props so it only displays on the appropriate pages

psealock commented on client/lib/nav-utils.js:
```diff
@@ -15,13 +16,28 @@ export const getAdminLink = path => {
+	currentQuery = currentQuery || parse( history.location.search.substring( 1 ) );
+	const queryString = stringifyQueryObject( Object.assign( currentQuery, query ) );
+	return `${ path }?${ queryString }`;
+};
```
Great abstraction, thank you. Here's another place it should be used in the date-picker

https://github.com/woocommerce/wc-admin/blob/b7dc5d19751bc6edf73b2a67d9fa3a6af531b327/client/components/date-picker/index.js#L70

ryelle commented:
In the most recent update to this PR, I've moved this parent component & all filter-related components into `component/filters`, and started to consolidate CSS. The new folder structure is below, I've left each filter in its own folder because each still has its own style.scss with specific styling overrides.

- components/filters/advanced/index.js - `<AdvancedFilters />`
- components/filters/date/index.js - `<DatePicker />`
- components/filters/filter/index.js - `<FilterPicker />`
- components/filters/index.js - `<ReportFilters />`

The base class names have updated, `woocommerce-filters__*`, `woocommerce-filters-date__*`, etc

I've also…

- removed FILTER_PARAM, we're just using query.filter now. I think constants will make it harder on third-party devs, esp those who are used to WP_Query. We can still document all query params, and I can do an audit of that in a week or so after filters work has settled somewhat
- Added two new functions, `getPath` & `getQuery`, which get path/query from the history object
- Updated the date-picker's path generation using the helper functions
- I moved the compareValues/presetValues to `lib/date` & imported from `lib/date`, rather than `lib/date` importing from the component

I'm sure moving files around collapsed the above comments, so here's what was left "unactioned":

- filters + filterPaths should merge into one config object – I don't want to do that in this PR since I'm already moving around a lot, I want my changes to stay superficial.

We still need to brainstorm maybe about how to handle passing around translated labels etc into our shared components - perhaps a good hangout topic? or are we just making it harder on ourselves and we should pass around objects of labels, [for example like post type labels?](https://developer.wordpress.org/reference/functions/get_post_type_labels/#source)

> The Tabular Component (p6riRB-3lZ-p2) will also manipulate filter parameters and I imagine may need to access some functions from within this folder.

That's what inspired this, I started working on the comparison filter (which is why [the framework for it is in place here](https://github.com/woocommerce/wc-admin/blob/412a1520766d99861cf53e80f3bf85aa1f65c071/client/components/filters/index.js#L26)) - so that's how I would see that section working.
